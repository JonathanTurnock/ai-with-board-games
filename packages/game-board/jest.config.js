module.exports = {
  preset: "ts-jest",
  testEnvironment: "node",
  rootDir: "./",
  testMatch: ["<rootDir>/test/**/*.test.ts", "<rootDir>/**/*.test.ts"],
};
