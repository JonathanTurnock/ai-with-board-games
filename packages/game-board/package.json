{
  "name": "game-board",
  "version": "0.1.0",
  "description": "> TODO: description",
  "main": "dist/index.js",
  "private": true,
  "scripts": {
    "clean": "rimraf results dist tsconfig.tsbuildinfo",
    "build:watch": "tsc-watch --outDir dist --rootDir src",
    "build": "tsc --outDir dist --rootDir src",
    "test": "jest --runInBand --passWithNoTests"
  },
  "dependencies": {
    "chalk": "^4.1.2",
    "excel-column-name": "^1.0.1",
    "lodash": "^4.17.21",
    "table": "^6.7.5"
  },
  "devDependencies": {
    "@types/jest": "^27.0.3",
    "@types/lodash": "^4.14.178",
    "jest": "^27.4.5",
    "rimraf": "^3.0.2",
    "ts-jest": "^27.1.2",
    "ts-node": "^10.4.0",
    "tsc-watch": "^4.6.0",
    "typescript": "^4.5.4"
  }
}
