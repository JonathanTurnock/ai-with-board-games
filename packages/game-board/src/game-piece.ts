import { black, blue, red } from "chalk";

export interface GamePieceProps {
  x: number;
  y: number;
  symbol: string;
  color: "blue" | "red";
}

export class GamePiece {
  constructor(private readonly props: GamePieceProps) {}

  get x() {
    return this.props.x;
  }

  get y() {
    return this.props.y;
  }

  toString(): string {
    switch (this.props.color) {
      case "red":
        return red(this.props.symbol);
      case "blue":
        return blue(this.props.symbol);
      default:
        return black(this.props.symbol);
    }
  }
}
