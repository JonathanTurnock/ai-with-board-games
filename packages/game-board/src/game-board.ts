import { range } from "lodash";
import { table } from "table";
import { GamePiece } from "./game-piece";
import excelColumnName from "excel-column-name";

export interface GameBoardProps {
  height: number;
  width: number;
}

export class GameBoard {
  private pieces: Record<string, GamePiece> = {};

  constructor(private readonly props: GameBoardProps) {}

  setPiece(piece: GamePiece) {
    this.pieces[`${piece.x},${piece.y}`] = piece;
  }

  render() {
    const board = range(this.props.height + 1).map((y) => [
      y > 0 ? y : " ",
      ...range(this.props.width).map((x) => {
        if (y === 0) return excelColumnName.intToExcelCol(x + 1);
        return this.pieces[`${x},${y - 1}`]?.toString() || " ";
      }),
    ]);
    console.log(
      table(board, {
        columnDefault: {
          paddingLeft: 2,
          paddingRight: 2,
        },
      })
    );
  }
}
