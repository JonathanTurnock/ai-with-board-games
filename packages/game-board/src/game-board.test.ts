import { GameBoard } from "./game-board";
import { GamePiece } from "./game-piece";

describe("GameBoard", () => {
  const board = new GameBoard({ height: 3, width: 3 });

  const piece1 = new GamePiece({ x: 0, y: 1, color: "blue", symbol: "x" });
  const piece2 = new GamePiece({ x: 0, y: 0, color: "blue", symbol: "x" });
  const piece3 = new GamePiece({ x: 1, y: 0, color: "red", symbol: "o" });

  board.setPiece(piece1);
  board.setPiece(piece2);
  board.setPiece(piece3);

  describe("render", () => {
    const consoleSpy = jest.spyOn(console, "log");

    it("should call console log", () => {
      board.render();
      expect(consoleSpy.mock.calls[0][0]).toMatchSnapshot();
    });
  });
});
