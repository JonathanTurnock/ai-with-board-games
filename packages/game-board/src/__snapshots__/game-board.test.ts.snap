// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GameBoard render should call console log 1`] = `
"╔═════╤═════╤═════╤═════╗
║     │  A  │  B  │  C  ║
╟─────┼─────┼─────┼─────╢
║  1  │  [34mx[39m  │  [31mo[39m  │     ║
╟─────┼─────┼─────┼─────╢
║  2  │  [34mx[39m  │     │     ║
╟─────┼─────┼─────┼─────╢
║  3  │     │     │     ║
╚═════╧═════╧═════╧═════╝
"
`;
